### PR TITLE
[expression] rewrite built-in string function: REPEAT

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -385,40 +385,37 @@ type repeatFunctionClass struct {
...
 	}
-	str, err := args[0].ToString()
-	if err != nil {
-		return d, err
+	num, isNull, err := b.args[1].EvalInt(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
 	}
+
 	ch := fmt.Sprintf("%v", str)
```
Comment (hanfei1991) [inline expression/builtin_string.go]: str itself is a string. You needn't assign it again.

File: expression/builtin_string.go [hunk 2]
```diff
@@ -385,40 +385,37 @@ type repeatFunctionClass struct {
 }
 
 func (c *repeatFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinRepeatSig{newBaseBuiltinFunc(args, ctx)}
+	resultTp := types.NewFieldType(mysql.TypeString)
```
Comment (shenli) [inline expression/builtin_string.go]: Please also set Flen.

File: expression/builtin_string.go (hunk 1 above)
Comment (zz-jason) [inline expression/builtin_string.go]: done

File: expression/builtin_string.go [hunk 3]
```diff
@@ -385,40 +385,37 @@ type repeatFunctionClass struct {
...
-		num = int(x.GetUint64())
-	}
 	if num < 1 {
-		d.SetString("")
-		return d, nil
+		return "", false, nil
 	}
-	d.SetString(strings.Repeat(ch, num))
-	return d, nil
+	return strings.Repeat(ch, int(num)), false, nil
```
Comment (zz-jason) [inline expression/builtin_string.go]: do as mysql: set `num` to max int, return null when the length of result exceed MAX_BLOB_WIDTH(16777216)

File: expression/builtin_string.go (hunk 2 above)
Comment (zz-jason) [inline expression/builtin_string.go]: done

File: expression/builtin_string.go [hunk 4]
```diff
@@ -385,40 +385,46 @@ type repeatFunctionClass struct {
...
+		return "", false, nil
 	}
-	d.SetString(strings.Repeat(ch, num))
-	return d, nil
+
+	if num > 0x7FFFFFFF {
+		num = 0x7FFFFFFF
+	}
+
+	if int64(len(str)) > int64(b.tp.Flen)/num {
```
Comment (hanfei1991) [inline expression/builtin_string.go]: Add test for this

File: expression/builtin_string.go [hunk 5]
```diff
@@ -385,40 +385,46 @@ type repeatFunctionClass struct {
 }
 
 func (c *repeatFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinRepeatSig{newBaseBuiltinFunc(args, ctx)}
+	tp := types.NewFieldType(mysql.TypeString)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: @zz-jason 
This type inferring is not complete.
For example,
``` sql
create table t(
	c_int int,
	c_double double,
    c_decimal decimal,
    c_binary binary(20),
    c_char char(20),
    c_varchar varchar(20),
    c_varbinary varbinary(20),
    c_blob blob,
    c_date date,
    c_datetime datetime,
    c_enum enum('a', 'b', 'c')
)
```
``` sql
select repeat(c_char, 2), repeat(c_varchar,2 ) , repeat(c_binary,2 ), repeat(c_varbinary,2 ), repeat(c_blob,2 ) from t;
```
In MySQL :

File: expression/builtin_string.go (hunk 4 above)
Comment (zz-jason) [inline expression/builtin_string.go]: done

**Comment by zz-jason** [comment]
@XuHuaiyu @hanfei1991 @shenli PTAL

File: expression/builtin_string_test.go [hunk 6]
```diff
@@ -287,6 +287,20 @@ func (s *testEvaluatorSuite) TestRepeat(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(v.GetString(), Equals, "aa")
 
+	args = []interface{}{"a", uint64(16777217)}
```
Comment (zz-jason) [inline expression/builtin_string_test.go]: done

File: expression/builtin_string.go [hunk 7]
```diff
@@ -383,40 +383,52 @@ type repeatFunctionClass struct {
...
+		return "", isNull, errors.Trace(err)
 	}
 	if num < 1 {
-		d.SetString("")
-		return d, nil
+		return "", false, nil
 	}
-	d.SetString(strings.Repeat(ch, num))
-	return d, nil
+	if num > 0x7FFFFFFF {
```
Comment (zz-jason) [inline expression/builtin_string.go]: use `math.MaxInt32` instead

File: expression/builtin_string.go [hunk 8]
```diff
@@ -383,40 +383,52 @@ type repeatFunctionClass struct {
...
-func (b *builtinRepeatSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
-	}
-	str, err := args[0].ToString()
-	if err != nil {
-		return d, err
+func (b *builtinRepeatSig) evalString(row []types.Datum) (d string, isNull bool, err error) {
+	str, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
```
Comment (zz-jason) [inline expression/builtin_string.go]: done

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**